### PR TITLE
refactor(eslint-remote-tester): convert to TS

### PR DESCRIPTION
**Pull request by MichaelDeBoey**

### CODE REVIEW

**Comment by MichaelDeBoey** [comment]
@AriPerkkio tests are fixed

**Comment by MichaelDeBoey** [comment]
@AriPerkkio tests are fixed

**Review by AriPerkkio — CHANGES_REQUESTED** [review]
Thanks for the improvements @MichaelDeBoey. 

I initially setup the repositories package as plain Javascript due to its simplicity. But I do like having it as Typescript so this is welcome. 👍 
This however brings some complexity to other setups. Currently weekly scans ran under `./ci` are failing with error:
```
Error: Cannot find module '/home/runner/work/eslint-remote-tester/eslint-remote-tester/ci/node_modules/eslint-remote-tester-repositories/dist/index.js'. Please verify that the package.json has a valid "main" entry
```
https://github.com/AriPerkkio/eslint-remote-tester/runs/2440214807?check_suite_focus=true

I believe before running those jobs the repositories package has to initialized and built. But I'm not sure what else is missing - I'm still running to that same error in local environment.

You should be able to run following in local environment once everything is setup properly:
```
cd ./ci
yarn log --config ./plugin-configs/eslint-core.config.js

./node_modules/.bin/eslint-remote-tester --config ./plugin-configs/eslint-core.config.js
```

**Comment by MichaelDeBoey** [comment]
@AriPerkkio I've rebased onto `master` & updated all workflows

**Comment by AriPerkkio** [comment]
CI jobs are now working OK. ✅

File: repositories/scripts/filter-unique-repos.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+const uniqueRepos = repos.filter(
+    (item, index, array) => array.indexOf(item) === index
+);
+
+console.log(
+    `Repositories ${repos.length}\nUnique repositories ${uniqueRepos.length}`
+);
+
+writeFileSync(
+    '../src/repositories.json',
```
Comment (AriPerkkio) [inline repositories/scripts/filter-unique-repos.ts:14]:
```suggestion
    './src/repositories.json',
```

File: repositories/package.json [hunk 2]
```diff
@@ -4,21 +4,35 @@
...
-    "main": "index.js",
+    "main": "dist/index.js",
+    "files": [
+        "dist"
+    ],
     "repository": {
         "type": "git",
-        "url": "https://github.com/AriPerkkio/eslint-remote-tester"
+        "url": "https://github.com/AriPerkkio/eslint-remote-tester",
+        "directory": "repositories"
```
Comment (AriPerkkio) [inline repositories/package.json:14]: I had no idea such field exists, good catch! 💯

**Comment by AriPerkkio** [comment]
https://github.com/AriPerkkio/eslint-remote-tester/blob/e6d5c47c2544f3395e5e437403e261b8f2bf767a/.github/workflows/ci.yml#L20-L22

Add `yarn build` for `./repositories` package here as well.
No idea how to add review comment on files which are not listed in changed. 😄

File: repositories/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import { pathIgnorePatterns } from './pathIgnorePatterns';
+import repositories from './repositories.json';
+
+const shuffle = <T>(array: T[]): T[] =>
```
[the remark this answers is not in the document]
Comment (MichaelDeBoey) [reply repositories/src/index.ts:4]: You can't make `functions` generic

File: repositories/scripts/filter-unique-repos.ts (hunk 1 above)
Comment (MichaelDeBoey) [reply repositories/scripts/filter-unique-repos.ts:14]: This path isn't correct

**Comment by MichaelDeBoey** [comment]
@AriPerkkio Updated al things requested

File: repositories/src/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (AriPerkkio) [reply repositories/src/index.ts:4]: Not sure what you mean. This is completely valid:

```ts
function shuffle<T>(array: T[]): T[] {
    return array
        .map(value => ({ value, order: Math.random() }))
        .sort((a, b) => a.order - b.order)
        .map(({ value }) => value);
}
```

[the remark this answers is not in the document]
Comment (MichaelDeBoey) [reply repositories/src/index.ts:4]: I also like to have implicit returns, as that makes things shorter

File: repositories/scripts/filter-unique-repos.ts (hunk 1 above)
Comment (AriPerkkio) [reply repositories/scripts/filter-unique-repos.ts:14]:
```
~/GIT/eslint-remote-tester/repositories$ yarn generate:repositories:unique
yarn run v1.22.5
$ ts-node ./scripts/filter-unique-repos.ts
Repositories 14496
Unique repositories 14496
Error: ENOENT: no such file or directory, open '../src/repositories.json'
```

**Review by AriPerkkio — APPROVED** [review]
Looks good, thanks for making the requested changes! 🎉

**Comment by G-Rath** [comment]
@AriPerkkio since the codebase is now in TypeScript (so there are type definitions) would you be open to supporting `eslint-remote-tester.config.ts` using `ts-node` similar to how it's done in Webpack & Jest (among others)?

I'm happy to have a go at implementing if you are (I've done it in a few projects now - it's typically a pretty straightforward & small diff)

**Pull request merged**